dbms/src/Functions/FunctionsRegexp.h

```diff
@@ -0,0 +1,1007 @@
+    using ParamStringNullableAndConst = Param<ParamString<true>, true>;
+
+    // Int type
+    using ParamIntNullableAndNotConst = Param<ParamInt<false>, true>;
+    using ParamIntNotNullableAndConst = Param<ParamInt<true>, false>;
+    using ParamIntNotNullableAndNotConst = Param<ParamInt<false>, false>;
+    using ParamIntNullableAndConst = Param<ParamInt<true>, true>;
+
+    enum class ParamType
+    {
+        StringNullableAndNotConst,
+        StringNotNullableAndConst,
+        StringNotNullableAndNotConst,
+        StringNullableAndConst,
+        IntNullableAndNotConst,
+        IntNotNullableAndConst,
+        IntNotNullableAndNotConst,
+        IntNullableAndConst
+    };
+
+    // default ParamString's ParamType should be ParamType::StringNotNullAndNotConst
+    ParamVariant(ColumnPtr col, size_t col_size, const StringRef & default_val)
+        : col_ptr(col)
+        , default_str(default_val)
+        , default_int(0)
+        , param(nullptr)
+    {
+        if (col_ptr != nullptr)
+        {
+            setParamStringTypeAndGenerateParam(col_size);
+        }
+        else
+        {
+            // This param is not provided by user, so we should use default value.
+            param = new ParamStringNotNullableAndConst(col_size, default_val);
+            param_type = ParamType::StringNotNullableAndConst;
+        }
+    }
+
+    // default ParamInt's ParamType should be ParamType::IntNotNullAndNotConst
+    ParamVariant(ColumnPtr col, size_t col_size [[maybe_unused]], Int64 default_val)
+        : col_ptr(col)
+        , default_str("", 0)
+        , default_int(default_val)
+        , param(nullptr)
+    {
+        // TODO implement it in next pr
+        throw Exception("Not implemented so far");
+    }
+
+    ~ParamVariant()
+    {
+        if (param != nullptr)
+        {
+            switch (param_type)
+            {
+// Enumerate cases that delete string param pointer
+#define M(NAME, pv_name, param_name, next_process)     \
+    case ParamType::NAME:                              \
+    {                                                  \
+        delete reinterpret_cast<Param##NAME *>(param); \
+        break;                                         \
+    }
+
+                // Expand the macro to enumerate string param cases
+                APPLY_FOR_PARAM_STRING_VARIANTS(M, placeholder1, placeholder2, placeholder3)
+#undef M
+
+// Enumerate cases that delete int param pointer
+#define M(NAME, pv_name, param_name, next_process)     \
+    case ParamType::NAME:                              \
+    {                                                  \
+        delete reinterpret_cast<Param##NAME *>(param); \
+        break;                                         \
+    }
+
+                // Expand the macro to enumerate int param cases
+                APPLY_FOR_PARAM_INT_VARIANTS(M, placeholder1, placeholder2, placeholder3)
+#undef M
```

Comment:
Seems these two macros are the same?  
BTW, `ENUMERATE_PARAM_VARIANT_CASES` can be used here.

Reply:
> Seems these two macros are the same? BTW, `ENUMERATE_PARAM_VARIANT_CASES` can be used here.

Done